### PR TITLE
RMB-96: count query RMB-175: query limit

### PR DESCRIPTION
**Pull request by wwelling**
Resolves https://issues.folio.org/browse/RMB-96 and https://issues.folio.org/browse/RMB-175.

This snapshot was tested against mod-users and mod-notify. I am now going to attempt upgrading some other modules and test.

### CODE REVIEW

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java [hunk 1]
```diff
@@ -1229,6 +1225,65 @@ private void doDelete(SQLConnection connection, boolean transactionMode, String
...
+          replyHandler.handle(Future.failedFuture(countQuery.cause()));
+        } else {
+
+          int total = countQuery.result().getInteger(0);
+
+          long countQueryTime = (System.nanoTime() - start);
+          StatsTracker.addStatElement(STATS_KEY + COUNT_STAT_METHOD, countQueryTime);
+          log.debug("timer: get " + q[1] + " (ns) " + countQueryTime);
+
+          // TODO: if total is 0 dont run query, requires building result with subset of arguments
```
[the remark this answers is not in the document]
Comment (wwelling) [reply domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: Knowing the count before the actually query, you can either opt out of the query entirely or adjust query accordingly.

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java [hunk 2]
```diff
@@ -1850,23 +1882,17 @@ public void join(JoinBy from, JoinBy to, String operation, String joinType, CQLW
    * @param setId
    * @return
    */
-  private <T> Results<T> processResult(ResultSet rs, Class<T> clazz, boolean count, boolean setId) {
+  private <T> Results<T> processResult(ResultSet rs, Class<T> clazz, Integer total, boolean setId) {
```
[the remark this answers is not in the document]
Comment (wwelling) [reply domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: Do we want to refactor this method  in this pull request?

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (wwelling) [reply domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: Also, if count is less than offset, query should not run.

**Comment by wwelling** [comment]
I am glad you looked at the commit history. That was intended to show the transition. If we prefer that not to be in the history, I will squash those commits.

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (julianladisch) [reply domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: This is not required for this pull request.

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java [hunk 3]
```diff
@@ -1158,67 +1176,47 @@ private void doDelete(SQLConnection connection, boolean transactionMode, String
       String fieldName, String where, boolean returnCount, boolean returnIdField, boolean setId,
       List<FacetField> facets, Handler<AsyncResult<Results<T>>> replyHandler) {
```
Comment (julianladisch) [inline domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: There should be more unit test coverage of doGet. You may drop "private" to make this function accessible for unit tests.
Unit tests help code review because they prove that the function works for basic input and for edge cases.

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (julianladisch) [reply domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: You may ignore id-jenkins' comment for this pull request.

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java [hunk 4]
```diff
@@ -1158,67 +1176,47 @@ private void doDelete(SQLConnection connection, boolean transactionMode, String
...
         if (!"null".equals(fieldName) && fieldName.contains("*")) {
-          //if we are requesting all fields (*) , then dont add the id field to the select
-          //this will return two id columns which will create ambiguity in facet queries
+          // if we are requesting all fields (*) , then dont add the id field to the select
+          // this will return two id columns which will create ambiguity in facet queries
           addIdField = "";
         }
 
-        String from2where = FROM + convertToPsqlStandard(tenantId) + "." + table + " " + where;
+        String[] q = new String[2];
```
Comment (julianladisch) [inline domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: Consider replacing q by two variables with better names, for example selectQuery and countQuery. Change function interface of processQueryWithCount accordingly.

Comment (wwelling) [reply domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: This refactoring makes sense. I will get these changes in.

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (wwelling) [reply domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: I am happy to write some unit tests. Although, I am not sure about making this method public. Exposing this function increase its use cases, which arguments are currently controlled by other methods within this class. Wouldn’t it be better to attempt having edge case tests be through the existing public methods?

[the remark this answers is not in the document]
Comment (julianladisch) [reply domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: Dropping "private" results in no explicit modifier and is package-private: https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html
Only classes in the same package can access it, including unit test classes in that package.

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java (hunk 3 above)
Comment (julianladisch) [reply domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: Dropping "private" results in no explicit modifier and is package-private: https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html
Only classes in the same package can access it, including unit test classes in that package.

File: domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (wwelling) [reply domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java]: Oh, I see. No identifier. Not necessarily public. That is reasonable. Honestly, I thought explicit access declaration was desired and no modifier was actually public. Thanks for clarifying.
